### DISCUSSION TITLE
How to Transfert Jason data inside a Rule to Domoticz (DualR3 with POW)?

### DISCUSSION THREAD

**Discussion by alain16** [arendst/Tasmota#20879] [Q&A]
Hello,

I want to manage POW of a DualR3 inside a Domoticz customevent.

Tasmota publish SENSOR state as :
`12:54:51.063 MQT: tele/DualR3-1/SENSOR = {"Time":"2024-03-05T12:54:51","Switch1":"ON","Switch2":"ON","ENERGY":
{"TotalStartTime":"2024-02-03T14:11:45","Total":0.895,"Yesterday":0.024,"Today":0.013,"Period":[0,0],"Power":[0,2],"ApparentPower":[0,8],"ReactivePower":[0,8],"Factor":[0.00,0.22],"Frequency":50,"Voltage":232,"Current":[0.000,0.037]}}`

So I do a Rule based on ENERGY like that:
`Rule1 ON ENERGY DO publish domoticz/in {"command":"customevent","event":"DualR3Action","data":"{\"idx\":2000,\"EnergyJason\":"%value%"}" } ENDON`

My objective is to send to Domoticz the jason part of "ENERGY":
`{"TotalStartTime":"2024-02-03T14:11:45","Total":0.895,"Yesterday":0.024,"Today":0.013,"Period":[0,0],"Power":[0,2],"ApparentPower":[0,8],"ReactivePower":[0,8],"Factor":[0.00,0.22],"Frequency":50,"Voltage":232,"Current":[0.000,0.037]}`

Unfortunatly, that's not worked and the traces are:
15:41:00.985 MQT: domoticz/in = {"command":"customevent","event":"DualR3Action","data":"{\"idx\":2000,\"EnergyJason\":}" }

How can I do to transfert the jason part of "ENERGY"?

Thanks for your help :)

**Comment by sfromis** [comment]
With the old rules feature, the answer is the same as in your other thread. Separate rules to pick up separate values, one by one.

However, as the Sonoff DualR3 has an ESP32 MCU, you can also create Berry code with full access to the JSON (not Jason) data from the sensor payload, and do what you like there.

**Reply by alain16** [reply]
Thanks sfromis, 
do you have a code sample in Berry close to what I want to do ?
Thanks, Alain

**Reply by sfromis** [reply]
Here's a rudimentary illustration of picking up data at TelePeriod, and doing publish. Of course pick the data and format you want instead. First line is to remove the old if you want to experiment....
``` 
tasmota.remove_rule('tele#ENERGY')
tasmota.add_rule('tele#ENERGY',
  def (data, trigger, payload)
    var voltage = data['Voltage']
    var power = data['Power']
    var message = f'v={voltage} p1={power[0]} p2={power[0]}'
    import mqtt
    mqtt.publish('mytopic', message)
  end)
```

**Reply by alain16** [reply]
Hi sfromis,

Thank you for your code. It helped me to better understand how Berry work and how powerful it is.
I will probably use class and Driver.

However, I'm testing f-strings and there is an syntax error that I don't understand.
Any idea of what happen?

tasmota.add_rule('tele#ENERGY',
  def (data, trigger, payload)
	var messageToAdd = f'\"data\":{data}'
	print("messageToAdd") print(messageToAdd) 
_It works well -> print is : {'Period': [0, 0], 'Frequency': 50, 'Current': [0, 0.036], 'Yesterday': 0.024, 'Voltage': 229, 'TotalStartTime': '2024-02-03T14:11:45', 'Today': 0.015, 'ReactivePower': [0, 8], 'Total': 0.968, 'ApparentPower': [0, 8], 'Factor': [0, 0.23], 'Power': [0, 2]}_

	 var message = '{\"command\":\"customevent\",\"event\":\"DualR3Action\",\"idx\":1000,\"data\":"Toto"}'
	print("message") print(message) 
_It works well -> print is : {"command":"customevent","event":"DualR3Action","idx":1000,"data":"Toto"}_

_but with:_
	 var message = f'{\"command\":\"customevent\",\"event\":\"DualR3Action\",\"idx\":1000,{messageToAdd}}' 
	print(message) print("message") 
_It generate BRY: Exception> 'syntax_error' - input:6: expected ')' before 'customevent'_

  end)

Thanks again for your help.
Alain

**Reply by sfromis** [reply]
When you use f-strings for formatting, you need to double each of `{` `}` as these are part of the syntax for both JSON and f-strings, meaning that when you double them in the latter, it will be passed on as just one.

However there's another way of doing it without carefully crafting the details of JSON, like this:
```
var messagemap = {'command':'customevent','event':'DualR3Action','idx':1000,'data':data}
import json
var message = json.dump(message)
```
This creates a map in Berry, and then converts the map to a string. While the notation for creating a map that way is similar to JSON, it is not JSON before using `json.dump` for that.
You could also instead add individual fields to the map, after being partially created, like:
```
messagemap['data']=data
```

**Reply by alain16** [reply]
Wow, the way you code is smart!
Thank you for your help to understand exception due to double { }
Have a nice weekend, it will be study for me ;)
